Jellyfin.Api/Controllers/NotificationsController.cs

```diff
@@ -0,0 +1,159 @@
+        public NotificationsController(INotificationManager notificationManager, IUserManager userManager)
+        {
+            _notificationManager = notificationManager;
+            _userManager = userManager;
+        }
+
+        /// <summary>
+        /// Gets a user's notifications.
+        /// </summary>
+        /// <param name="userId">The user's ID.</param>
+        /// <param name="isRead">An optional filter by notification read state.</param>
+        /// <param name="startIndex">The optional index to start at. All notifications with a lower index will be omitted from the results.</param>
+        /// <param name="limit">An optional limit on the number of notifications returned.</param>
+        /// <response code="200">Notifications returned.</response>
+        /// <returns>An <see cref="OkResult"/> containing a list of notifications.</returns>
+        [HttpGet("{UserID}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public ActionResult<NotificationResultDto> GetNotifications(
+            [FromRoute] string userId,
+            [FromQuery] bool? isRead,
+            [FromQuery] int? startIndex,
+            [FromQuery] int? limit)
+        {
+            return new NotificationResultDto();
+        }
+
+        /// <summary>
+        /// Gets a user's notification summary.
+        /// </summary>
+        /// <param name="userId">The user's ID.</param>
+        /// <response code="200">Summary of user's notifications returned.</response>
+        /// <returns>An <cref see="OkResult"/> containing a summary of the users notifications.</returns>
+        [HttpGet("{UserID}/Summary")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public ActionResult<NotificationsSummaryDto> GetNotificationsSummary(
+            [FromRoute] string userId)
+        {
+            return new NotificationsSummaryDto();
+        }
+
+        /// <summary>
+        /// Gets notification types.
+        /// </summary>
+        /// <response code="200">All notification types returned.</response>
+        /// <returns>An <cref see="OkResult"/> containing a list of all notification types.</returns>
+        [HttpGet("Types")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public ActionResult<IEnumerable<NotificationTypeInfo>> GetNotificationTypes()
+        {
+            return _notificationManager.GetNotificationTypes();
+        }
+
+        /// <summary>
+        /// Gets notification services.
+        /// </summary>
+        /// <response>All notification services returned.</response>
+        /// <returns>An <cref see="OkResult"/> containing a list of all notification services.</returns>
+        [HttpGet("Services")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public ActionResult<IEnumerable<NameIdPair>> GetNotificationServices()
+        {
+            return _notificationManager.GetNotificationServices().ToList();
+        }
```

Comment:
```suggestion
        public IEnumerable<NameIdPair> GetNotificationServices()
        {
            return _notificationManager.GetNotificationServices();
        }
```

Reply:
This is what I originally had, I changed it according to [this](https://github.com/jellyfin/jellyfin/issues/2872#issuecomment-617845103).

> Endpoints should always return ActionResult or ActionResult<T>, which will make things more consistent and also more clear about the response codes being returned

Reply:
I think we discovered that `IEnumerable` or `IAsyncEnumerable` should be returned instead if possible, according to the docs. Hopefully this is the last change :)

Reply:
Updated!

Reply:
@crobibero I'm missing where in the docs it is recommended to not use `ActionResult<T>` in this case. What if we have an endpoint returning `IEnumerable` that might also need to return a 404 in some cases?

Reply:
@mark-monteiro Ok it seems I got a bit lost following the discussion over in #2872. My understanding _was_ that `ActionResult<IEnumerable<T>>` should be returned in case of other potential return paths, and that we would just return it in all cases for consistency. But based on @crobibero's comment I thought the decision was to return just `IEnumerable<T>` when there weren't alternate return paths.

Reply:
For the case when returning an `IEnumerable` with a possible `NotFound` status:

```
public ActionResult<IEnumerable<User>> GetUser(bool id)
{
	if (id)
	{
		return NotFound();
	}

	return Ok(_userManager.Users);
}
```

According to this [stackoverflow answer](https://stackoverflow.com/a/59843524/11704592) returning in this manner is also non-blocking

Reply:
> My understanding _was_ that `ActionResult<IEnumerable<T>>` should be returned in case of other potential return paths, and that we would just return it in all cases for consistency.

This is what I thought as well. Looking back at the way I worded my comment though I see now that I wasn't super clear about what to do for `IEnumerable<>`. @crobibero I do think removing `ToList()` is correct here as you suggested, but perhaps we should still use `ActionResult<T>` to maintain consistency as agreed upon in the other thread?

@ZadenRB Sorry we're sorting this out in the middle of your PR 🤦

Reply:
No worries at all. Unfortunately though the solution you suggested here of removing `ToList()` won't work (it's what I originally tried to do). According to [the docs]()
> C# doesn't support implicit cast operators on interfaces. Consequently, conversion of the interface to a concrete type is necessary to use `ActionResult<T>`.
So we either have to make the return type be just `IEnumerable<T>` and not use `ToList()`, or stick with `ActionResult<IEnumerable<T>>`, maintain consistency, and use `ToList()` or `ToListAsync()`.

Reply:
We could only return `IEnumerable` directly is when filtering in the controller is required, such as: 
https://github.com/jellyfin/jellyfin/pull/2929/files#diff-735f348ef92662378c348030d074120aR60

though this could just be converted to a `Where` clause.

Reply:
This appears to work, you just need to use the `Ok()` method, which accepts any object. This is consistent with our convention of using `ActionResult<T>` and doesn't require calling `ToList()`

```cs
public ActionResult<IEnumerable<NameIdPair>> GetNotificationServices()
{
    return Ok(_notificationManager.GetNotificationServices());
}
```

Reply:
There are two real options:

IAsyncEnumerable
IEnumerable
ActionResult<IAsyncEnumerable>
ActionResult<IEnumerable>
ActionResult<T>
ActionResult

OR

ActionResult<IAsyncEnumerable>
ActionResult<IEnumerable>
ActionResult<T>
ActionResult


if we go with the second option it forces some actions to be formatted as:
```
[HttpGet]
[ProducesResponseType(StatusCodes.Status200OK)]
public ActionResult<IEnumerable<IScheduledTaskWorker>> GetTasks(
	[FromQuery] bool? isHidden = false,
	[FromQuery] bool? isEnabled = false)
{
	IEnumerable<IScheduledTaskWorker> tasks = _taskManager.ScheduledTasks
		.OrderBy(o => o.Name)
		.Where(task =>
		{
			if (task.ScheduledTask is IConfigurableScheduledTask scheduledTask)
			{
				if (isHidden.HasValue && isHidden.Value != scheduledTask.IsHidden)
				{
					return false;
				}

				if (isEnabled.HasValue && isEnabled.Value != scheduledTask.IsEnabled)
				{
					return false;
				}
			}

			return true;
		});

	return Ok(tasks);
}
```

instead of 

```
[HttpGet]
[ProducesResponseType(StatusCodes.Status200OK)]
public IEnumerable<IScheduledTaskWorker> GetTasks(
	[FromQuery] bool? isHidden = false,
	[FromQuery] bool? isEnabled = false)
{
	IEnumerable<IScheduledTaskWorker> tasks = _taskManager.ScheduledTasks.OrderBy(o => o.Name);

	foreach (var task in tasks)
	{
		if (task.ScheduledTask is IConfigurableScheduledTask scheduledTask)
		{
			if (isHidden.HasValue && isHidden.Value != scheduledTask.IsHidden)
			{
				continue;
			}

			if (isEnabled.HasValue && isEnabled.Value != scheduledTask.IsEnabled)
			{
				continue;
			}
		}

		yield return task;
	}
}
```

I believe both functions will have roughly the same performance, so it doesn't matter which we choose to me

Reply:
I'd say let's _prefer_ using `ActionResult<IEnumerable<T>>` to maintain consistency with other endpoints as much as possible, but also allow returning `IEnumerable<>` directly where it makes sense.

I don't think we don't need to be super strict on this, especially if there is existing code using `yield return`, or that syntax makes more sense to use than `.Where()` for a specific method. I can update my guidance comment in the other issue if we're in agreement.

Reply:
The `yield return` is from my PR for `ScheduledTasks`, and I doubt there are any existing methods that use that structure. 
I agree with your statement, prefer using `ActionResult<IEnumerable<T>>`, lets get these PRs rolling!

Reply:
Also agreed on preferring `ActionResult<IEnumerable<T>>` for consistency, if their functionality & performance are the same.

Reply:
Great, the documentation comment has been updated to clarify this 👍